core/src/main/java/org/infinispan/container/EntryFactory.java

```diff
@@ -1,68 +1,110 @@
 package org.infinispan.container;
 
 import org.infinispan.atomic.Delta;
+import org.infinispan.commands.VisitableCommand;
 import org.infinispan.container.entries.CacheEntry;
-import org.infinispan.container.entries.MVCCEntry;
+import org.infinispan.container.entries.NullCacheEntry;
 import org.infinispan.context.InvocationContext;
+import org.infinispan.interceptors.impl.EntryWrappingInterceptor;
+import org.infinispan.interceptors.locking.ClusteringDependentLogic;
+import org.infinispan.topology.CacheTopology;
 
 /**
  * A factory for constructing {@link org.infinispan.container.entries.MVCCEntry} instances for use in the {@link org.infinispan.context.InvocationContext}.
  * Implementations of this interface would typically wrap an internal {@link org.infinispan.container.entries.CacheEntry}
- * with an {@link org.infinispan.container.entries.MVCCEntry}, optionally acquiring the necessary locks via the
- * {@link org.infinispan.util.concurrent.locks.LockManager}.
+ * with an {@link org.infinispan.container.entries.MVCCEntry}.
+ *
+ * <h3>Expected wrapping patterns</h3>
+ *
+ * {@link EntryWrappingInterceptor} checks {@link CacheTopology#getReadConsistentHash()} and if this node is an owner
+ * of the key, it loads the entry from {@link DataContainer}. On the way back through interceptor stack, the entry is
+ * committed from {@link EntryWrappingInterceptor} through {@link ClusteringDependentLogic} which checks
+ * {@link CacheTopology#getWriteConsistentHash()}.
+ * <p>
+ * Entry being wrapped is a prerequisite for the command to run and therefore commit the entry, but it's not up to
+ * {@link EntryWrappingInterceptor} to make sure the entry is always wrapped - the invariant is
+ *       (can read locally) => (entry is wrapped).
```

Comment:
I meant this as invariant for interceptors below, but you're right, I am not clear in my use of fancy words :) Anyway, the other direction does not hold, entry may be wrapped because:
a) it is an L1 entry (by EWI)
b) the command is about to write the entry (by distribution interceptor) because this node is owner according to writeCH (but not readCH)
c) this is a transactional cache (by dist. int. after remoteGet)